core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -258,6 +268,314 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+//     returned function environment is populated with the following metadata objects:
+//     - .defined:  signature method
+//     - .target:   signature target
+//     - .Generic:
+//     - .Method:   method definition
+//     - .Methods:  function call
+//     - e1:        arg1
+//     - e2:        arg2
+//
+//     otherwise only e1 and e2.
+  
+    if (method.getGroup().equals("generic")) { /*  && method.getDistance() == 0 */
+      return context.evaluate(new FunctionCall(selectedMethods.get(0).getFunction(), args));
+    } else {
+      SEXP objDefined = generateDefinedObjectMetadata(context, method);
+      SEXP generic = generateObjectGenericMetadata(opName);
+    
+      Map<Symbol, SEXP> metadata = new HashMap<>();
+      metadata.put(Symbol.get(".defined"), objDefined);
+      metadata.put(Symbol.get(".Generic"), generic);
+      metadata.put(Symbol.get(".Method"), method.getFunction());
+      metadata.put(Symbol.get(".Methods"), Symbol.get(".Primitive(\"" + opName +"\")"));
+      metadata.put(Symbol.get(".target"), objDefined);
+    
+      return ClosureDispatcher.apply(context,rho, new FunctionCall(method.getFunction(), args), (Closure) method.getFunction(), args, metadata);
+    }
+  }
+  
+  private static SEXP generateObjectGenericMetadata(String opName) {
+    SEXP generic = Symbol.get(opName);
+    generic.setAttribute("package", StringVector.valueOf("base"));
+    return generic;
+  }
+  
+  private static SEXP generateDefinedObjectMetadata(@Current Context context, SelectedMethod method) {
+    StringVector.Builder objDefinedSig = new StringVector.Builder();
+    String[] objDefinedClasses = method.getInputSignature().split("#");
+    for(int i = 0; i < objDefinedClasses.length; i++) {
+      objDefinedSig.add(objDefinedClasses[i]);
+    }
+    SEXP objDefined = Symbol.get(deparse(context, objDefinedSig.build(), 256, false, 0, 1));
+    objDefined.setAttribute(".Names", ((Closure) method.getFunction()).getFormals().getNames());
+    
+    StringVector.Builder objDefinedPkg = new StringVector.Builder();
+    for(int i = 0; i < objDefinedClasses.length; i++) {
+      Environment environment = context.getGlobalEnvironment();
+      SEXP classS4Object = environment.findVariable(context, Symbol.get(".__C__" + objDefinedClasses[i]));
+      objDefinedPkg.add(classS4Object.getAttribute(Symbol.get("package")).asString());
+    }
+    objDefined.setAttribute("package", objDefinedPkg.build());
+    
+    StringVector.Builder sv2 = new StringVector.Builder();
+    sv2.add("signature");
+    SEXP classSignature = sv2.build();
+    StringVector.Builder sv3 = new StringVector.Builder();
+    sv3.add("methods");
+    classSignature.setAttribute("package", sv3.build());
+    objDefined.setAttribute("class", classSignature);
+    return objDefined;
+  }
+  
+  private static Environment findMethodEnvironment(@Current Context context, String opName) {
```

Comment:
The `@Current` annotation is for builtins and internals, definitely does not belong on private methods.